src/main/scala/is/hail/variant/MatrixTable.scala

```diff
@@ -750,41 +750,13 @@ class MatrixTable(val hc: HailContext, val ast: MatrixIR) extends JoinAnnotator
   }
 
   def annotateSamplesTable(kt: Table, vdsKey: java.util.ArrayList[String],
-    root: String, expr: String, product: Boolean): MatrixTable =
-    annotateSamplesTable(kt, if (vdsKey != null) vdsKey.asScala else null, root, expr, product)
+    root: String, product: Boolean): MatrixTable =
+    annotateSamplesTable(kt, if (vdsKey != null) vdsKey.asScala else null, root, product)
 
   def annotateSamplesTable(kt: Table, vdsKey: Seq[String] = null,
-    root: String = null, expr: String = null, product: Boolean = false): MatrixTable = {
+    root: String = null, product: Boolean = false): MatrixTable = {
 
-    if (root == null && expr == null || root != null && expr != null)
-      fatal("method `annotateSamplesTable' requires one of `root' or 'expr', but not both")
-
-    var (joinSignature, f): (Type, Annotation => Annotation) = kt.valueSignature.size match {
-      case 0 => (TBoolean(), _ != null)
-      case 1 => (kt.valueSignature.fields.head.typ, x => if (x != null) x.asInstanceOf[Row].get(0) else null)
-      case _ => (kt.valueSignature, identity[Annotation])
-    }
-
-    if (product) {
-      joinSignature = if (joinSignature.isInstanceOf[TBoolean]) TInt32() else TArray(joinSignature)
-      f = if (kt.valueSignature.size == 0)
-        _.asInstanceOf[IndexedSeq[_]].length
-      else {
-        val g = f
-        _.asInstanceOf[IndexedSeq[_]].map(g)
-      }
-    }
-
-    val (finalType, inserter): (TStruct, (Annotation, Annotation) => Annotation) = {
-      val (t, ins) = if (expr != null) {
-        val ec = EvalContext(Map(
-          "sa" -> (0, colType),
-          "table" -> (1, joinSignature)))
-        Annotation.buildInserter(expr, colType, ec, Annotation.SAMPLE_HEAD)
-      } else insertSA(joinSignature, root)
-
-      (t, (a: Annotation, toIns: Annotation) => ins(a, f(toIns)))
-    }
+    val (finalType, inserter) = colType.structInsert(kt.valueSignature, List(root))
```

Comment:
This all looks great except I'm confused about why the to-be-inserted-struct-type does not depend on the value of `product`. Can you explain that to me? The tests pass, so I must misunderstand something (or we need more exhaustive tests).

Reply:
this definitely looks like a bug.

Reply:
let's get some tests that expose it as a part of this PR too!

Reply:
yup, will do that first! I'm not sure we can do it in Python yet, but definitely in Scala.

Reply:
or maybe we can with view_join_x...